x/restaking/keeper/end_blocker_test.go

```diff
@@ -314,3 +316,97 @@ func (suite *KeeperTestSuite) TestKeeper_CompleteMatureUnbondingDelegations() {
 		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestGasConsumption_UndelegateVsEndBlockProcessing() {
+	for _, numDenomsPerDelegation := range []int{1, 10, 30, 100, 500} {
+		suite.Run(fmt.Sprintf("with%dDenoms", numDenomsPerDelegation), func() {
+			ctx, _ := suite.ctx.CacheContext()
+
+			ctx = ctx.
+				WithBlockHeight(10).
+				WithBlockTime(time.Date(2024, 1, 1, 12, 0, 0, 0, time.UTC))
+
+			// Set the unbonding time to 7 days
+			params, err := suite.k.GetParams(ctx)
+			suite.Require().NoError(err)
+			params.UnbondingTime = 7 * 24 * time.Hour
+			err = suite.k.SetParams(ctx, params)
+			suite.Require().NoError(err)
+
+			// Create a delegator address
+			delegator := "cosmos167x6ehhple8gwz5ezy9x0464jltvdpzl6qfdt4"
+
+			// Generate denoms to delegate
+			denoms := make([]string, numDenomsPerDelegation)
+			for i := range denoms {
+				denoms[i] = fmt.Sprintf("denom%d", i)
+			}
+
+			// Create a service to delegate to
+			const serviceID = 1
+			serviceAddress := servicestypes.GetServiceAddress(serviceID).String()
+			err = suite.sk.SaveService(ctx, servicestypes.Service{
+				ID:              serviceID,
+				Status:          servicestypes.SERVICE_STATUS_ACTIVE,
+				Address:         serviceAddress,
+				Tokens:          sdk.NewCoins(),
+				DelegatorShares: sdk.NewDecCoins(),
+			})
+			suite.Require().NoError(err)
+
+			// Fund the delegator account with sufficient balance
+			initialBalance := sdk.NewCoins()
+			for _, denom := range denoms {
+				initialBalance = initialBalance.Add(sdk.NewInt64Coin(denom, 1000_000000))
+			}
+			suite.fundAccount(ctx, delegator, initialBalance)
+
+			// Prepare the total amount to delegate
+			delAmt := sdk.NewCoins()
+			for _, denom := range denoms {
+				delAmt = delAmt.Add(sdk.NewInt64Coin(denom, 100_000000))
+			}
+
+			// Delegate multiple denominations to the service using MsgServer
+			//
+			// NOTE: We don't include this as part of the gas comparison because delegations can
+			// be cheaply accumulated over arbitrary amounts of time and undelegated in a batch in a single block.
+			// The attack vector hinges on this burst of undelegations to pack >1 block worth of unmetered gas
+			// into a single block of gas consumption.
+			msgServer := keeper.NewMsgServer(suite.k)
```

Comment:
I think we should reset the gas meter here, to make sure we do not count the setup inside the undelegation gas cost. This can be done by passing a new gas meter to the `WithGasMeter`  method of the context: 

```go
ctx = ctx.WithGasMeter(types.NewUnlimitedGasMeter())
```

Reply:
I agree that resetting the gas meter before operations we want to measure would make the code easier to read.

FYI since we call `undelegationGasStart := ctx.GasMeter().GasConsumed()` right before we call `UndelegateService`, we're not counting the setup. It's like calling `started := time.Now()` before an operation and calling `elapsed := time.Since(started)` after it.